File: ios/Sources/MemoryInfoPluginPlugin/MemoryInfoPluginPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(MemoryInfoPluginPlugin)
public class MemoryInfoPluginPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "MemoryInfoPluginPlugin"
    public let jsName = "MemoryInfoPlugin"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "echo", returnType: CAPPluginReturnPromise)
    ]
    private let implementation = MemoryInfoPlugin()

    @objc func echo(_ call: CAPPluginCall) {
        let value = call.getString("value") ?? ""
        call.resolve([
            "value": implementation.echo(value)
        ])
    }

    @objc func getRAMUsage(_ call: CAPPluginCall) {
        let pageSize = vm_kernel_page_size
        var stats = vm_statistics64()
        var count = mach_msg_type_number_t(MemoryLayout<vm_statistics64_data_t>.size / MemoryLayout<integer_t>.size)
        let hostPort = mach_host_self()
        
        let result = withUnsafeMutablePointer(to: &stats) {
            $0.withMemoryRebound(to: integer_t.self, capacity: Int(count)) {
                host_statistics64(hostPort, HOST_VM_INFO64, $0, &count)
            }
        }
        
        guard result == KERN_SUCCESS else {
            call.reject("Failed to fetch VM stats")
            return
        }
        
        let totalMemory = ProcessInfo.processInfo.physicalMemory
        let freeMemory = UInt64(stats.free_count) * UInt64(pageSize)
        let usedMemory = totalMemory - freeMemory
        
        let totalRAMGB = Double(totalMemory) / 1_073_741_824.0
        let availableRAMGB = Double(freeMemory) / 1_073_741_824.0
        let usedRAMGB = Double(usedMemory) / 1_073_741_824.0

        call.resolve([
            "totalRAM": String(format: "%.2f", totalRAMGB),
            "availableRAM": String(format: "%.2f", availableRAMGB),
            "usedRAM": String(format: "%.2f", usedRAMGB)
        ])
    }
}

File: android/src/main/java/com/memory/info/MemoryInfoPluginPlugin.java
package com.memory.info;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import android.app.ActivityManager;
import android.content.Context;

@CapacitorPlugin(name = "MemoryInfoPlugin")
public class MemoryInfoPluginPlugin extends Plugin {

    private MemoryInfoPlugin implementation = new MemoryInfoPlugin();

    @PluginMethod
    public void echo(PluginCall call) {
        String value = call.getString("value");

        JSObject ret = new JSObject();
        ret.put("value", implementation.echo(value));
        call.resolve(ret);
    }

    @PluginMethod
    public void getRAMUsage(PluginCall call) {
        ActivityManager activityManager = (ActivityManager) getContext().getSystemService(Context.ACTIVITY_SERVICE);
        ActivityManager.MemoryInfo memoryInfo = new ActivityManager.MemoryInfo();
        activityManager.getMemoryInfo(memoryInfo);

        long totalRAM = memoryInfo.totalMem;
        long availableRAM = memoryInfo.availMem;
        long usedRAM = totalRAM - availableRAM;

        double totalRAMGB = totalRAM / 1_073_741_824.0;
        double availableRAMGB = availableRAM / 1_073_741_824.0;
        double usedRAMGB = usedRAM / 1_073_741_824.0;

        JSObject ret = new JSObject();
        ret.put("totalRAM", String.format("%.2f", totalRAMGB));
        ret.put("availableRAM", String.format("%.2f", availableRAMGB));
        ret.put("usedRAM", String.format("%.2f", usedRAMGB));

        call.resolve(ret);
    }
}
